src/main/java/parsing/Parser.java

```diff
@@ -81,10 +121,22 @@ public NodeGraph parse(final String filename, NodeGraph graph) {
                         graph.addEdge(from, to);
                         line = in.readLine();
                     }
-                } else {
+               } else if (line.startsWith("S") && !newCache) {
```

Comment:
Im thinking you can take this branch and the previous together and make a nested if-statement to check on newCache and do the required thing in either case.